peps/pep-0734.rst

```diff
@@ -0,0 +1,798 @@
+interpreter has a distinct object wrapping that data.  For all other
+shareable types, a strict copy or proxy is made such that the
+corresponding objects continue to match exactly.  In cases where
+the underlying data is complex but must be copied (e.g. ``tuple``),
+the data is serialized as efficiently as possible.
+
+Shareable objects must be specifically supported internally
+by the Python runtime.  However, there is no restriction against
+adding support for more types later.
+
+Here's the initial list of supported objects:
+
+* ``str``
+* ``bytes``
+* ``int``
+* ``float``
+* ``bool`` (``True``/``False``)
+* ``None``
+* ``tuple`` (only with shareable items)
+* ``Queue``
+* ``memoryview`` (underlying buffer actually shared)
+
+Note that the last two on the list, queues and ``memoryview``, are
+technically mutable data types, whereas the rest are not.  When any
+interpreters share mutable data there is always a risk of data races.
+Cross-interpreter safety, including thread-safety, is a fundamental
+feature of queues.
+
+However, ``memoryview`` does not have any native accommodations.
+The user is responsible for managing thread safety, whether passing
+a token back and forth through a queue to indicate safety
+(see `Synchronization`_), or by assigning sub-range exclusivity
+to individual interpreters.
+
+Regarding queues, the primary situation for sharing is where an
+interpreter is using ``prepare___main__()`` to provide the target
+interpreter with their means of further communication.
+
+Finally, a reminder: for some types the actual object is shared,
+whereas for others only the underlying data (or even a copy or proxy)
+is shared.  Regardless, the guarantee of "shareable" objects is that
+corresponding objects in different interpreters will always strictly
+match each other.
+
+Synchronization
+---------------
+
+There are situations where two interpreters should be synchronized.
+That may involve sharing a resource, worker management, or preserving
+sequential consistency.
+
+In threaded programming the typical synchronization primitives are
+types like mutexes.  The ``threading`` module exposes several.
+However, interpreters cannot share objects which means they cannot
+share ``threading.Lock`` objects.
+
+The ``interpreters`` module does not provide any such dedicated
+synchronization primitives.  Instead, ``Queue`` objects provide
+everything we need.
+
+For example, if there's a shared resource that needs managed
+access then a queue may be used to manage it, where the interpreters
+pass an object around to indicate who can use the resource::
+
+   import interpreters
+   from mymodule import load_big_data, check_data
+
+   numworkers = 10
+   control = interpreters.create_queue()
+   data = memoryview(load_big_data())
+
+   def worker():
+       interp = interpreters.create()
+       interp.prepare___main__(control=control, data=data)
+       interp.exec("""if True:
```

Comment:
Looks like we're missing the end of the triple-quoted strings in these two examples?

It's more obvious from the preview:

https://pep-previews--3523.org.readthedocs.build/pep-0734/#synchronization